custom_components/resmed_myair/client/rest_client.py

```diff
@@ -0,0 +1,704 @@
+    # Gives sessionToken once verified
+    "mfa_url": "https://{okta_url}/api/v1/authn/factors/{email_factor_id}/verify?rememberDevice=true",
+    # Authorization endpoint to send sessionToken to in order to get 'code'.
+    "authorize_url": "https://{okta_url}/oauth2/{auth_server_id}/v1/authorize",
+    # The endpoint that the 'code' is sent to get an access token
+    "token_url": "https://{okta_url}/oauth2/{auth_server_id}/v1/token",
+    # Checks the access token to see if it is still active or not
+    "introspect_url": "https://{okta_url}/oauth2/{auth_server_id}/v1/introspect",
+    # Uses the access token to return ResMed user info
+    "userinfo_url": "https://{okta_url}/oauth2/{auth_server_id}/v1/userinfo",
+}
+
+class RESTClient(MyAirClient):
+    """
+    myAir uses oauth on Okta and AWS AppSync GraphQL
+    """
+
+    def __init__(self, config: MyAirConfig, session: ClientSession) -> None:
+        _LOGGER.debug(
+            f"[RESTClient init] config: {async_redact_data(config._asdict(), KEYS_TO_REDACT)}"
+        )
+        self._config: MyAirConfig = config
+        self._session: ClientSession = session
+        self._json_headers: dict[str, Any] = {
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        }
+        self._country_code: str | None  = None
+        self._access_token: str | None  = None
+        self._id_token: str | None  = None
+        self._state_token: str | None  = None
+        self._session_token: str | None  = None
+        self._cookie_dt: str | None = self._config.device_token
+        self._cookie_sid: str | None  = None
+        self._uses_mfa: bool = False
+        if self._config.region == REGION_NA:
+            self._region_config: dict[str, Any] = NA_CONFIG
+        else:
+            self._region_config = EU_CONFIG
+        self._email_factor_id: str = self._region_config["email_factor_id"]
+        self._mfa_url: str = OAUTH_URLS["mfa_url"].format(
+            okta_url=self._region_config["okta_url"],
+            email_factor_id=self._email_factor_id,
+        )
+
+    @property
+    def device_token(self) -> str | None:
+        return self._cookie_dt
+
+    @property
+    def _cookies(self) -> dict[str, Any]:
+        cookies: dict[str, Any] = {}
+        if self._cookie_dt:
+            cookies["DT"] = self._cookie_dt
+        if self._cookie_sid:
+            cookies["sid"] = self._cookie_sid
+        # _LOGGER.debug(f"[cookies] returning cookies: {cookies}")
+        return cookies
+
+    async def connect(self, initial: bool | None = False) -> str:
+        if self._cookie_dt is None:
+            await self._get_initial_dt()
+        if self._cookie_dt is None and self._uses_mfa:
+            _LOGGER.warning("Device Token isn't set. This will require frequent reauthentication.")
+        if self._access_token and await self._is_access_token_active():
+            return AUTHN_SUCCESS
+        _LOGGER.info("Starting Authentication")
```

Comment:
should this be debug?

Reply:
No, the Device Token cookie is the main thing that Okta uses to remember the previous MFA authentication. If the Device Token isn't set, every time the access token expires (2 hours) it will require MFA again. I am now getting an initial device token to mitigate this so this shouldn't happen much.